docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,119 @@
+
+1. keeping them in some vault inside a controller implementation
+2. not exposing them to any microservices in a controller except for the party which decrypts from the vault and re-encrypts for a particular device
+3. not exposing them to all microservices on the device and accidentally storing them in files or putting them in logs
+4. potentially restricting it by policy on the device so that in a defense-in-depth implementation only the downloader can access datastore credentials, only nim can access WiFi credentials (basically least privilege approach)
+5. MiTM proxies and server-side load balancers which terminate TLS should not be able to see this information
+
+## Overview
+
+The approach is to encrypt such credential and other private information using a per-device key, and doing this encryption before assigning it to the fields which get protobuf encoded. Thus other microservices in a controller and the device will pass them along as any protobuf to golang data structure, but the content will be encrypted.
+
+This means that the protobuf API will have new fields which are the encrypted fields (with the old username/password fields etc still in place as we transition away from only protecting this using TLS).
+
+And it requires that the controller and the device can create an encryption key, and a mechanism to re-key from either end. The shared encryption key is created using the standard ECDH based on a ECC key pair on the controller and an ECC  key pair on the device for this purpose, where each end shares their certificate containing the ECC public key.
+
+## ECDH key agreement
+
+[Elliptic Curve Diffie Hellman(ECDH)](https://en.wikipedia.org/wiki/Elliptic-curve_Diffie%E2%80%93Hellman) Key Exchange is used for establishing a shared secret key based on the ECC certificates. ECDH does itself not provide authentication of the two endpoints, but the fact that the certificates are exchanged between the device and the controller using TLS plus object signing provide for that authentication.
+
+The shared secret is fed through SHA-256 as a key derivation function per [NIST SP 800-56A Rev. 3](https://csrc.nist.gov/publications/detail/sp/800-56a/rev-3/final) and [FIPS PUB 180-4](https://csrc.nist.gov/publications/detail/fips/180/4/final).
+
+The derived key is used an [AES 256 CTS key](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard) with an initialization vector passed from the controller as part of the device configuration message. This key and IV are then used for encrypting the sensitive information between the controller and the device.
+
+The ECC certificates used in the exchange are:
+
+1. A certificate that the device creates during first install, specifically for ECDH operations.
```

Comment:
This is distinct from the usual device key? Where is that stored, and is it more protected than the regular device key?

Reply:
Stored in the TPM. It needs to have some encryption bit set in the certificate I think. Checking with @cshari-zededa

Reply:
From @cshari-zededa : The TPM key used for device cert is a Signing key, it does not support Decryption. For ECDH to work, TPM expects a key with Decryption set and Signing unset.

That means for devices in the field, we need to create one more key for ECDH exchange
that results in one more cert to manage